### DISCUSSION TITLE
Custom outliner icons

### DISCUSSION THREAD

**Discussion by dj-mcg** [Autodesk/maya-usd#2499] [Ideas]
Hey folks,

Our users have expressed an interest in specifying custom outliner icons for various nodes. Some of the icons would depend on logic or data other than prim type. For now, we are patching this behavior in, but I had thought it might be useful for other users if we added the ability to register a `GetIconOverride` function (or something similarly named) that `UsdUIInfoHandler::treeViewIcon` could call prior to falling back to the default type-based icons. 

I'm not sure if providing arbitrary paths to `baseIcon` is something UFE/Maya would support right now, so I wanted to open up a discussion about what's possible now and what we may make available down the road. Thanks!

Dan

**Comment by santosg87** [comment]
@dj-mcg  would you be able to give us a little more info on what you are trying to achieve? we have a couple of possibilities here, but we would like to understand a bit better so we can make a more educated discussion. @ppt-adsk any ideas on this?

**Reply by ppt-adsk** [reply]
Hi @dj-mcg, are you talking about a base icon that can change dynamically?  To our knowledge this is not supported by Maya, only dynamically changing badges on the base icon, and this may be tricky, as the number (4) of badge positions is limited, and used by maya-usd to represent data model characteristics (e.g. variants), so this might be conflicting.  Is this something you think is worth exploring?

Icon file paths are looked up in resources, then if not found are looked up in paths described by the XBMLANGPATH environment variable.  Is this sufficient for your purposes?

Thanks to @seando-adsk for input.

**Reply by ppt-adsk** [reply]
Hi @dj-mcg, after further investigation by @seando-adsk (thanks!), we believe we actually can support dynamically changing base icon pixmaps, as long as a refresh of the Outliner is triggered externally.  The change would need to be done here:
https://github.com/Autodesk/maya-usd/blob/c8291ec838debacc71de5994ec3dcb845a6a52ea/lib/mayaUsd/ufe/UsdUIInfoHandler.cpp#L162
so that whatever dynamic criterion is desired would affect the Ufe::UIInfoHandler::Icon::baseIcon data member.  In quick proof of concept testing, we have this working.  Here is code to refresh all Outliners, again courtesy of @seando-adsk:
```
// Refresh all the Outliner panels
string $outPanels[] =` getPanel -type "outlinerPanel"`;
string $outPanel;
for ($outPanel in $outPanels) {
    outlinerEditor -e -rfs $outPanel;
}
```
If you get a proof of concept working, we can figure out how to deploy this non-invasively, so that the code in UsdUIInfoHandler.cpp can remain unchanged.  Apologies for the delay in getting the right answer, hopefully this will be more helpful.

**Reply by dj-mcg** [reply]
Cool, thanks for all the info! We'll give this a shot once everyone is back from SIGGRAPH!

**Reply by jossh-h** [reply]
Hi @dj-mcg , we are also looking at applying custom icons, were you or any of your team able to get this working?

**Reply by dj-mcg** [reply]
Hey @jossh-h - We unfortunately haven't had a chance to try this yet. We have a hacked/patched solution that got us what we need for now, but I'd still like to try this approach, especially if it reduces/eliminates any merge conflicts that we occasionally encounter. Let us know if you are successful with this approach!

**Reply by jossh-h** [reply]
Sounds good @dj-mcg , will post back here if we get something working.
Cheers
